pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -208,10 +346,31 @@ private RoutingTable getRoutingTable(String tableName, TableType tableType, long
         CalciteSqlCompiler.compileToBrokerRequest("SELECT * FROM " + tableNameWithType), requestId);
   }
 
+  // --------------------------------------------------------------------------
+  // Partitioned leaf stage assignment
+  // --------------------------------------------------------------------------
   private void assignWorkersToPartitionedLeafFragment(DispatchablePlanMetadata metadata,
-      DispatchablePlanContext context, String partitionKey, int numPartitions, int partitionParallelism) {
+      DispatchablePlanContext context, String partitionKey, Map<String, String> tableOptions) {
+    // when partition key exist, we assign workers for leaf-stage in partitioned fashion.
+
+    String numPartitionsStr = tableOptions.get(PinotHintOptions.TableHintOptions.PARTITION_SIZE);
+    Preconditions.checkState(numPartitionsStr != null, "'%s' must be provided for partition key: %s",
+        PinotHintOptions.TableHintOptions.PARTITION_SIZE, partitionKey);
+    int numPartitions = Integer.parseInt(numPartitionsStr);
+    Preconditions.checkState(numPartitions > 0, "'%s' must be positive, got: %s",
+        PinotHintOptions.TableHintOptions.PARTITION_SIZE, numPartitions);
+
+    String partitionFunction = tableOptions.getOrDefault(PinotHintOptions.TableHintOptions.PARTITION_FUNCTION,
```

Comment:
Don't add default. In most cases this is `murmur` but we should not assume it

Reply:
this would be backward-incompatible, all existing queries with table hints without hash function will fail

Reply:
Understood. Currently we assume all table options using the same partition function. Putting `Murmur` as the default might be safer

Reply:
done